File: ios/TacticalMaps/Calibration/PDFTiler.swift
import Foundation
import UIKit
import PDFKit
import MapKit

/// Bakes a calibrated `PDFMapSource` into an offline MBTiles raster pyramid
/// **on-device** — no desktop GDAL step. Swift mirror of the Android `PdfTiler`:
/// maps each Web-Mercator XYZ tile's WGS84 box back to PDF user-space points via
/// the inverse calibration affine, renders that region with PDFKit, and writes
/// PNG tiles via `MBTilesWriter`.
enum PDFTiler {

    struct Progress { let done: Int; let total: Int }

    /// Returns the written .mbtiles URL, or nil on failure / no calibration.
    static func generate(source: PDFMapSource,
                         progress: @escaping (Progress) -> Void) -> URL? {
        guard let transform = source.placementTransform,
              let inverse = transform.inverted(),
              let region = source.coverage,
              let doc = PDFDocument(url: source.url),
              let page = doc.page(at: 0) else { return nil }

        let mediaBox = page.bounds(for: .mediaBox)
        let minLat = region.center.latitude - region.span.latitudeDelta / 2
        let maxLat = region.center.latitude + region.span.latitudeDelta / 2
        let minLon = region.center.longitude - region.span.longitudeDelta / 2
        let maxLon = region.center.longitude + region.span.longitudeDelta / 2

        let (minZoom, maxZoom) = zoomRange(minLat: minLat, maxLat: maxLat,
                                           minLon: minLon, maxLon: maxLon)

        func range(_ z: Int) -> WebMercatorTiles.Range {
            WebMercatorTiles.tileRange(minLat: minLat, maxLat: maxLat,
                                       minLon: minLon, maxLon: maxLon, z: z)
        }

        var total = 0
        for z in minZoom...maxZoom { total += range(z).count }
        guard total > 0 else { return nil }

        let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
        let dir = appSupport.appendingPathComponent("offline_tiles", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true,
                                                  attributes: [.protectionKey: FileProtectionType.complete])
        let outURL = dir.appendingPathComponent("tacmap-\(Int(Date().timeIntervalSince1970)).mbtiles")
        // Bake into a .partial temp and only publish to the real path on full
        // success, so an interrupted run can't leave a truncated file that later
        // loads as a "valid" (but incomplete) basemap.
        let tmpURL = outURL.appendingPathExtension("partial")

        guard let writer = MBTilesWriter(path: tmpURL.path) else { return nil }
        writer.writeMetadata(name: source.displayName, minZoom: minZoom, maxZoom: maxZoom,
                             minLon: minLon, minLat: minLat, maxLon: maxLon, maxLat: maxLat)

        let fmt = UIGraphicsImageRendererFormat()
        fmt.scale = 1
        fmt.opaque = true
        let imageRenderer = UIGraphicsImageRenderer(size: CGSize(width: 256, height: 256), format: fmt)

        var done = 0
        for z in minZoom...maxZoom {
            // Honour cancellation (Cancel button) — stop and clean up the temp.
            if Task.isCancelled {
                writer.close()
                try? FileManager.default.removeItem(at: tmpURL)
                return nil
            }
            let r = range(z)
            guard r.minX <= r.maxX, r.minY <= r.maxY else { continue }
            writer.begin()
            for tx in r.minX...r.maxX {
                for ty in r.minY...r.maxY {
                    let box = WebMercatorTiles.tileBounds(z, tx, ty)
                    if pdfRect(inverse: inverse, box: box, mediaBox: mediaBox) != nil,
                       let data = renderTile(imageRenderer, page: page,
                                             z: z, x: tx, y: ty,
                                             inverse: inverse, mediaBox: mediaBox) {
                        writer.putTile(z: z, x: tx, y: ty, data: data)
                    }
                    done += 1
                    if done % 16 == 0 { progress(Progress(done: done, total: total)) }
                }
            }
            writer.commit()
        }
        progress(Progress(done: total, total: total))
        writer.close()

        // A tile/metadata write failed mid-bake (e.g. disk full) — don't pass a
        // half-baked file off as a complete basemap.
        guard !writer.hadError else {
            try? FileManager.default.removeItem(at: tmpURL)
            return nil
        }
        // Atomic publish.
        do {
            try? FileManager.default.removeItem(at: outURL)
            try FileManager.default.moveItem(at: tmpURL, to: outURL)
        } catch {
            try? FileManager.default.removeItem(at: tmpURL)
            return nil
        }
        return outURL
    }

    /// Map a tile's WGS84 box to a PDF user-space rect (y-up). Used only as the
    /// off-page skip gate — a tile whose PDF rect doesn't touch the page is not
    /// baked at all. The actual render (`renderTile`) re-derives per-strip rects.
    private static func pdfRect(inverse: AffineTransform2D,
                                box: WebMercatorTiles.Box,
                                mediaBox: CGRect) -> CGRect? {
        // inverse.apply(CGPoint(x: lon, y: lat)) -> (.longitude = pdfX, .latitude = pdfY)
        let pts = [
            inverse.apply(CGPoint(x: box.west, y: box.south)),
            inverse.apply(CGPoint(x: box.east, y: box.south)),
            inverse.apply(CGPoint(x: box.east, y: box.north)),
            inverse.apply(CGPoint(x: box.west, y: box.north))
        ]
        let xs = pts.map { $0.longitude }
        let ys = pts.map { $0.latitude }
        guard let left = xs.min(), let right = xs.max(),
              let bottom = ys.min(), let top = ys.max() else { return nil }
        let rect = CGRect(x: left, y: bottom, width: right - left, height: top - bottom)
        guard rect.width > 0.01, rect.height > 0.01, rect.intersects(mediaBox) else { return nil }
        return rect
    }

    /// PDF-space bounding rect of the sub-band whose *Web-Mercator* tile-Y runs
    /// `[yTop, yBottom]` (tile units). Longitude is linear in both Mercator and
    /// the affine, so only the latitude edges are re-derived per strip.
    private static func stripRect(inverse: AffineTransform2D,
                                  west: Double, east: Double,
                                  yTop: Double, yBottom: Double, z: Int) -> CGRect? {
        let north = WebMercatorTiles.tileYToLat(yTop, z)
        let south = WebMercatorTiles.tileYToLat(yBottom, z)
        let pts = [
            inverse.apply(CGPoint(x: west, y: south)),
            inverse.apply(CGPoint(x: east, y: south)),
            inverse.apply(CGPoint(x: east, y: north)),
            inverse.apply(CGPoint(x: west, y: north))
        ]
        let xs = pts.map { $0.longitude }, ys = pts.map { $0.latitude }
        guard let left = xs.min(), let right = xs.max(),
              let bottom = ys.min(), let top = ys.max() else { return nil }
        return CGRect(x: left, y: bottom, width: right - left, height: top - bottom)
    }

    /// Render one 256×256 tile north-up. To stay Web-Mercator-correct we split the
    /// tile into horizontal strips that are each linear in tile-Y (Mercator) and
    /// map each strip through the calibration affine at its *true* latitude edges.
    /// A single linear scale over the whole tile would warp any tile spanning more
    /// than ~2° of latitude (large sheets at low zoom); strips reduce the residual
    /// to sub-pixel. Small high-zoom tiles span <¼° → one strip → zero extra cost.
    private static func renderTile(_ renderer: UIGraphicsImageRenderer,
                                   page: PDFPage, z: Int, x: Int, y: Int,
                                   inverse: AffineTransform2D, mediaBox: CGRect) -> Data? {
        let size: CGFloat = 256
        let box = WebMercatorTiles.tileBounds(z, x, y)
        // One strip per ≤¼° of latitude, capped at 16 (residual warp ≪ 1 px).
        let strips = max(1, min(16, Int(ceil((box.north - box.south) / 0.25))))

        let image = renderer.image { rctx in
            let ctx = rctx.cgContext
            UIColor.white.setFill()
            ctx.fill(CGRect(x: 0, y: 0, width: size, height: size))

            for i in 0..<strips {
                // Strip i covers tile pixel rows [pixelTop, pixelBottom] from the
                // top (north). Rounded integer edges so adjacent bands abut with
                // no hairline seam.
                let pixelTop = CGFloat((Double(i) * Double(size) / Double(strips)).rounded())
                let pixelBottom = CGFloat((Double(i + 1) * Double(size) / Double(strips)).rounded())
                let bandH = pixelBottom - pixelTop
                guard bandH > 0 else { continue }
                let yTop = Double(y) + Double(i) / Double(strips)
                let yBottom = Double(y) + Double(i + 1) / Double(strips)
                guard let r = stripRect(inverse: inverse, west: box.west, east: box.east,
                                        yTop: yTop, yBottom: yBottom, z: z),
                      r.width > 0.01, r.height > 0.01, r.intersects(mediaBox) else { continue }
                ctx.saveGState()
                ctx.clip(to: CGRect(x: 0, y: pixelTop, width: size, height: bandH))
                // Map strip rect (y-up PDF) onto pixel band [pixelTop, pixelBottom] (y-down).
                ctx.translateBy(x: 0, y: pixelTop)
                ctx.scaleBy(x: size / r.width, y: -bandH / r.height)
                ctx.translateBy(x: -r.minX, y: -r.maxY)
                page.draw(with: .mediaBox, to: ctx)
                ctx.restoreGState()
            }
        }
        return image.pngData()
    }

    /// Min zoom (coverage ~fits one tile) accumulating up to a tile budget so a
    /// huge sheet can't generate forever. Resolution-agnostic.
    private static func zoomRange(minLat: Double, maxLat: Double,
                                  minLon: Double, maxLon: Double) -> (Int, Int) {
        func count(_ z: Int) -> Int {
            WebMercatorTiles.tileRange(minLat: minLat, maxLat: maxLat,
                                       minLon: minLon, maxLon: maxLon, z: z).count
        }
        var minZoom = 0
        for z in 0...19 {
            if count(z) <= 4 { minZoom = z } else { break }
        }
        var maxZoom = minZoom
        var total = 0
        for z in minZoom...19 {
            total += count(z)
            if total > 2500 { break }
            maxZoom = z
        }
        return (minZoom, maxZoom)
    }
}

File: android/app/src/main/java/com/tacmap/calibration/PdfTiler.kt
package com.tacmap.calibration

import android.content.Context
import android.graphics.Bitmap
import android.graphics.RectF
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.withContext
import java.io.ByteArrayOutputStream
import java.io.File
import kotlin.math.abs
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.log2
import kotlin.math.roundToInt

/**
 * Bakes a calibrated [PdfMapSource] into an offline MBTiles raster pyramid
 * on-device, no desktop GDAL needed. Maps each Web-Mercator XYZ tile's WGS84
 * box back into PDF pixel space via the inverse calibration affine, renders
 * that region with [PdfPageRenderer], writes PNG tiles via [MBTilesWriter].
 *
 * Any GeoPDF or calibrated scanned sheet becomes a true offline basemap
 * with zero desktop tooling.
 */
object PdfTiler {

    data class Progress(val done: Int, val total: Int)

    /** Returns the written .mbtiles path, or null on failure / nothing to do. */
    suspend fun generate(
        context: Context,
        source: PdfMapSource,
        onProgress: (Progress) -> Unit
    ): String? = withContext(Dispatchers.IO) {
        val info = source.pageInfo ?: return@withContext null
        val transform = when (val c = source.calibration) {
            is Calibration.Fiduciaries -> c.transform
            is Calibration.Parsed -> c.transform
            null -> return@withContext null
        }
        val inverse = transform.inverted() ?: return@withContext null
        val coverage = source.coverage ?: return@withContext null

        val (minZoom, maxZoom) = zoomRange(info, coverage)
        var total = 0
        for (z in minZoom..maxZoom) total += WebMercatorTiles.tileRange(coverage, z).count
        if (total == 0) return@withContext null

        val dir = File(context.filesDir, "offline_tiles").apply { mkdirs() }
        val outFile = File(dir, "tacmap-${System.currentTimeMillis()}.mbtiles")
        // Bake into a .partial temp, only publish on full success. An
        // interrupted run can't leave a truncated file that later loads
        // as a "valid" but incomplete basemap.
        val tmpFile = File(dir, "${outFile.name}.partial")
        val outPath = outFile.absolutePath
        val writer = MBTilesWriter.create(tmpFile.absolutePath) ?: return@withContext null
        writer.writeMetadata(
            name = source.displayName,
            minZoom = minZoom,
            maxZoom = maxZoom,
            bounds = coverage
        )

        var done = 0
        try {
            for (z in minZoom..maxZoom) {
                // honour cancellation (Cancel button) - bail instead of baking
                // every zoom level after the user backed out
                ensureActive()
                val range = WebMercatorTiles.tileRange(coverage, z)
                writer.beginBatch()
                for (tx in range.minX..range.maxX) {
                    for (ty in range.minY..range.maxY) {
                        val box = WebMercatorTiles.tileBounds(z, tx, ty)
                        // whole-tile off-page gate; strips re-derive per-band rects
                        if (pdfPixelRect(inverse, box, info) != null) {
                            val strips = buildStrips(inverse, box, z, ty, info)
                            val png = if (strips.isEmpty()) null else runCatching {
                                val bmp = PdfPageRenderer.renderFirstPageStrips(
                                    context, source.uri, strips, TILE, TILE
                                )
                                val bytes = ByteArrayOutputStream().use {
                                    bmp.compress(Bitmap.CompressFormat.PNG, 100, it); it.toByteArray()
                                }
                                bmp.recycle()
                                bytes
                            }.getOrNull()
                            if (png != null) writer.putTile(z, tx, ty, png)
                        }
                        done++
                        if (done % 16 == 0) onProgress(Progress(done, total))
                    }
                }
                writer.commitBatch()
            }
            onProgress(Progress(total, total))
            writer.close()
            // tile/metadata write failed mid-bake (e.g. disk full), don't
            // pass a half-baked file off as complete basemap
            if (writer.hadError) {
                tmpFile.delete()
                null
            } else {
                File(outPath).delete()
                if (tmpFile.renameTo(File(outPath))) outPath else { tmpFile.delete(); null }
            }
        } catch (c: kotlinx.coroutines.CancellationException) {
            // cancelled by user, clean up temp and propagate so the caller's
            // coroutine ends without reporting a failure
            writer.close()
            tmpFile.delete()
            throw c
        } catch (_: Throwable) {
            writer.close()
            tmpFile.delete()
            null
        }
    }

    /** Map tile's WGS84 box to clamped PDF-pixel rect, null if off-page. */
    private fun pdfPixelRect(
        inverse: AffineTransform2D,
        box: Wgs84Bounds,
        info: PdfPageInfo
    ): RectF? = pageRectForBand(
        inverse,
        west = box.southwest.longitude, east = box.northeast.longitude,
        north = box.northeast.latitude, south = box.southwest.latitude,
        info = info
    )

    /**
     * Split a tile into <=0.25 deg latitude horizontal strips, each mapped
     * through the calibration affine at its true latitude edges (recovered
     * from Mercator tile-Y via [WebMercatorTiles.tileYToLat]). A single
     * region+FILL over the whole tile would warp anything spanning >~2 deg
     * of latitude (large sheets at low zoom); strips cut residual to
     * sub-pixel. Small high-zoom tiles yield one strip, same cost.
     */
    private fun buildStrips(
        inverse: AffineTransform2D,
        box: Wgs84Bounds,
        z: Int,
        ty: Int,
        info: PdfPageInfo
    ): List<PdfPageRenderer.RenderStrip> {
        val north = box.northeast.latitude
        val south = box.southwest.latitude
        val west = box.southwest.longitude
        val east = box.northeast.longitude
        val strips = ceil((north - south) / 0.25).toInt().coerceIn(1, 16)
        val out = ArrayList<PdfPageRenderer.RenderStrip>(strips)
        for (i in 0 until strips) {
            // Rounded integer pixel bands so adjacent strips abut with no seam.
            val topPx = (i.toDouble() * TILE / strips).roundToInt()
            val botPx = ((i + 1).toDouble() * TILE / strips).roundToInt()
            if (botPx <= topPx) continue
            // Strip i covers tile-Y [ty+i/strips, ty+(i+1)/strips]; convert those
            // Mercator edges back to their true latitudes.
            val bandNorth = WebMercatorTiles.tileYToLat(ty + i.toDouble() / strips, z)
            val bandSouth = WebMercatorTiles.tileYToLat(ty + (i + 1).toDouble() / strips, z)
            val pageRect = pageRectForBand(inverse, west, east, bandNorth, bandSouth, info)
                ?: continue // strip fully off-page, leave it white
            out += PdfPageRenderer.RenderStrip(
                pageRect = pageRect,
                dest = RectF(0f, topPx.toFloat(), TILE.toFloat(), botPx.toFloat())
            )
        }
        return out
    }

    /** PDF-pixel rect (y-down) for one lat/lon band, null if no meaningful
     *  overlap with the page. */
    private fun pageRectForBand(
        inverse: AffineTransform2D,
        west: Double, east: Double, north: Double, south: Double,
        info: PdfPageInfo
    ): RectF? {
        // inverse.apply(lon, lat) -> Wgs84Coordinate with (pdfX as longitude, pdfY as latitude)
        val corners = listOf(
            inverse.apply(west, south),
            inverse.apply(east, south),
            inverse.apply(east, north),
            inverse.apply(west, north)
        )
        val xs = corners.map { it.longitude }
        val ys = corners.map { it.latitude }
        val left = xs.min(); val right = xs.max()
        val w = info.pageWidth.toDouble(); val h = info.pageHeight.toDouble()
        // Calibration affine is in PDF user space (y-up, origin bottom-left,
        // per GeoPDF georeferencing dicts) but PdfRenderer draws y-down
        // (origin top-left). Flip Y here - feeding y-up coords straight to
        // the renderer mirrored every tile north-for-south. iOS does the
        // same flip in its CoreGraphics context.
        val top = h - ys.max()
        val bottom = h - ys.min()
        // skip bands that don't overlap the page at all
        if (right <= 0.0 || left >= w || bottom <= 0.0 || top >= h) return null
        // need non-trivial on-page overlap, don't emit all-margin bands
        val ovW = minOf(w, right) - maxOf(0.0, left)
        val ovH = minOf(h, bottom) - maxOf(0.0, top)
        if (ovW < 0.5 || ovH < 0.5) return null
        // return FULL (unclamped) rect: renderer maps it onto the tile band
        // so on-page content keeps true scale/position, off-page stays white
        return RectF(left.toFloat(), top.toFloat(), right.toFloat(), bottom.toFloat())
    }

    /** Min zoom (coverage roughly fits one tile) up to native-res max, capped
     *  by total-tile budget so a huge sheet can't generate forever. */
    private fun zoomRange(info: PdfPageInfo, coverage: Wgs84Bounds): Pair<Int, Int> {
        val lonSpan = abs(coverage.longitudeSpan).coerceAtLeast(1e-9)
        val pxPerDeg = info.pageWidth / lonSpan
        var maxZoom = floor(log2(pxPerDeg * 360.0 / TILE)).toInt().coerceIn(1, 19)

        var minZoom = 0
        for (z in 0..maxZoom) {
            if (WebMercatorTiles.tileRange(coverage, z).count <= 4) minZoom = z else break
        }
        minZoom = minZoom.coerceAtMost(maxZoom)

        while (maxZoom > minZoom) {
            var total = 0
            for (z in minZoom..maxZoom) total += WebMercatorTiles.tileRange(coverage, z).count
            if (total <= MAX_TILES) break
            maxZoom--
        }
        return minZoom to maxZoom
    }

    private const val TILE = 256
    private const val MAX_TILES = 2500
}
